File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0743_Network_Delay_Time.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.*;

public class Question_0743_Network_Delay_Time {
    private Map<Integer, List<Pair<Integer, Integer>>> adj = new HashMap<>();

    private static class Pair<A, B> {
        public final A key;
        public final B value;

        public Pair(A key, B value) {
            this.key = key;
            this.value = value;
        }

        public static <A, B> Pair<A, B> of(A a, B b) {
            return new Pair<A, B>(a, b);
        }

        public A getKey() {
            return key;
        }

        public B getValue() {
            return value;
        }
    }

    public int networkDelayTime(int[][] times, int n, int k) {
        for (int[] time : times) {
            int source = time[0];
            int dest = time[1];
            int travelTime = time[2];
            if (!adj.containsKey(source)) {
                adj.put(source, new ArrayList<>());
            }
            adj.get(source).add(Pair.of(travelTime, dest));
        }
        int[] signalReceivedAt = new int[n + 1];
        Arrays.fill(signalReceivedAt, Integer.MAX_VALUE);

        dijkstra(signalReceivedAt, k);

        int res = Integer.MIN_VALUE;
        for (int i = 1; i <= n; i++) {
            res = Math.max(res, signalReceivedAt[i]);
        }
        return res == Integer.MAX_VALUE ? -1 : res;
    }

    private void dijkstra(int[] signalReceivedAt, int source) {
        Queue<Pair<Integer, Integer>> pq = new PriorityQueue<>(Comparator.comparing(Pair::getKey));
        pq.add(Pair.of(0, source));
        signalReceivedAt[source] = 0;
        while (!pq.isEmpty()) {
            Pair<Integer, Integer> topPair = pq.remove();
            int currNode = topPair.getValue();
            int currNodeTime = topPair.getKey();
            if (currNodeTime > signalReceivedAt[currNode]) {
                continue;
            }
            if (!adj.containsKey(currNode)) {
                continue;
            }
            for (Pair<Integer, Integer> edge : adj.get(currNode)) {
                int time = edge.getKey();
                int neighborNode = edge.getValue();
                if (signalReceivedAt[neighborNode] > currNodeTime + time) {
                    signalReceivedAt[neighborNode] = currNodeTime + time;
                    pq.add(Pair.of(signalReceivedAt[neighborNode], neighborNode));
                }
            }
        }
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0743_Network_Delay_Time.cc
#include <limits.h>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int networkDelayTime(vector<vector<int>>& times, int n, int k) {
    for (vector<int>& t : times) {
      m[t[0]].push_back({t[2], t[1]});
    }
    vector<int> dist(n + 1, INT_MAX);
    dijkstra(dist, k);
    int res = INT_MIN;
    for (int i = 1; i <= n; i++) {
      res = max(res, dist[i]);
    }
    return res == INT_MAX ? -1 : res;
  }

 private:
  unordered_map<int, vector<pair<int, int>>> m;

  void dijkstra(vector<int>& dist, int source) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, source});
    dist[source] = 0;
    while (!pq.empty()) {
      pair<int, int> p = pq.top();
      pq.pop();
      int t = p.first;
      int s = p.second;
      if (t > dist[s]) {
        continue;
      }
      for (pair<int, int>& pp : m[s]) {
        int tt = pp.first;
        int ss = pp.second;
        if (dist[ss] > t + tt) {
          dist[ss] = t + tt;
          pq.push({t + tt, ss});
        }
      }
    }
  }
};
